### ISSUE TITLE
Feature Request: MiniMax-Text-01 model

### ISSUE THREAD

**Issue by Kreijstal** [ggml-org/llama.cpp#11290]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Please add support for minimax-text-01 model https://huggingface.co/MiniMaxAI/MiniMax-Text-01
https://github.com/MiniMax-AI/MiniMax-01

### Motivation

We need to add support for the latest models! Performs almost as good as deepseek v3. But has 4 million tokens context.

### Possible Implementation

It's a MoE model.

**Comment by ehartford** [comment]
Very interested in this model!

**Comment by ggerganov** [comment]
> I guess that proper support of this model would require some redesign of KV cache. The problem is that layers in MiniMax-Text-01 can either use linear lightning attention (a single kv matrix per layer is cached in this case) or regular transformers attention that caches separate key and value vectors. I need to think about it some more.

Yup, it's unfeasible to keep trying to fit all variants of the attention into the existing KV cache code. I am hoping that after the refactoring of #11213 , we will be able to implement custom attention mechanism for use cases like these.

**Comment by fairydreaming** [comment]
I noticed a problem with the model "eating" some words when asked to repeat text (Q5_K_M quant). Can someone with more RAM (like 512GB or 1TB) test this model with my branch? I'm not sure if the model is very sensitive to quantization or there is some other problem. The full prompt is:

```
<beginning_of_sentence>user name=user
Repeat this text: "The different accidents of life are not so changeable as the feelings of human nature. I had worked hard for nearly two years, for the sole purpose of infusing life into an inanimate body. For this I had deprived myself of rest and health. I had desired it with an ardour that far exceeded moderation; but now that I had finished, the beauty of the dream vanished, and breathless horror and disgust filled my heart."<end_of_sentence>
<beginning_of_sentence>ai name=assistant

```

while the model answer is:

```
The different accidents of life are not so changeable as the feelings human nature. I had worked hard for nearly two years, for the sole purpose of infusing life into an inanimate body. For this I had deprived myself rest and health. I had desired it with an ardour that far exceeded moderation; but now that I had finished, the beauty of the dream vanished, and breathless horror and disgust filled my heart.<end_of_sentence>
```

There is one missing "of" in front of "human nature" and another "of" in front of "rest and health". Sometimes it eats "and" instead or both. A hungry model. I ran it with temp 0.01.

I'm curious if it happens also on f16 or Q8_0 quantization.

**Comment by ehartford** [comment]
I have 1tb ram, I can try it

**Comment by fairydreaming** [comment]
I found about `llama_sbatch::split_equal`, so my branch now supports inference of multiple token sequences with llama-server. Prompt caching should be disabled for now, it doesn't work correctly. Run the server with `--jinja` to use model prompt template.

**Comment by Nondzu** [comment]
@fairydreaming tested your branch with Q5_K_M.  On my setup I see some missing "of".
Tested on AMD EPYC and 768G RAM.
Can you share your full command test to run? 
Building Q8 and will do test tomorrow...

`build: 4532 (1e74c4d9) with gcc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0 for x86_64-linux-gnu`

Q5_K_M:
```
> <beginning_of_sentence>user name=user  
Repeat this text: "The different accidents of life are not so changeable as the feelings of human nature. I had worked hard for nearly two years, for the sole purpose of infusing life into an inanimate body. For this I had deprived myself of rest and health. I had desired it with an ardour that far exceeded moderation; but now that I had finished, the beauty of the dream vanished, and breathless horror and disgust filled my heart."<end_of_sentence>  
<beginning_of_sentence>ai name=assistant  

"The different accidents life are not so changeable as the feelings human nature. I had worked hard for nearly two years, for the sole purpose of infusing life an inanimate body. For I had deprived myself of rest and health. I had desired it with an ardour that far exceeded moderation; but now that I had finished the beauty of the dream vanished, and breathless horror and disgust filled my heart."

 ```

**Comment by fairydreaming** [comment]
> [@fairydreaming](https://github.com/fairydreaming) tested your branch with Q5_K_M. On my setup I see some missing "of". Tested on AMD EPYC and 768G RAM. Can you share your full command test to run? Building Q8 and will do test tomorrow...

That would be helpful, thanks. Regarding the command line I can't access the workstation now, will add that later.

**Comment by Nondzu** [comment]
file format = GGUF V3 (latest)
file type   = Q8_0
file size   = 451.36 GiB (8.50 BPW) 

Full log: 

[minimax-q8.log](https://github.com/user-attachments/files/18549031/minimax-q8.log)

`compared by chatgpt`

### Summary of Rounds and Missing Words

Across the four rounds, the text provided by the user was analyzed for differences in word usage. Here's a concise summary of the missing words in each round and how they evolved:

---

### **Round 1**:
#### Missing Words:
1. **"of"** (in "The different accidents life are not so changeable as the feelings of human nature").
2. **"of"** (in "For this I had deprived myself rest health").
3. **"and"** (in "For this I had deprived myself rest health").
4. **"of"** (in "the beauty the dream vanished").
5. **"and"** (in "the beauty the dream vanished breathless horror").

---

### **Round 2**:
#### Missing Words:
1. **"of"** (in "as the feelings human nature").
2. **"of"** (in "for the sole purpose infusing life").

---

### **Round 3**:
- **No Missing Words**: The AI response matched the original text completely.

---

### **Round 4**:
- **No Missing Words**: The AI response was identical to the original text.

---

### **Summary of All Missing Words**:
From **Rounds 1 and 2**, the following words were missing:
1. **"of"** (five occurrences in total across both rounds).
2. **"and"** (two occurrences in Round 1).

In **Rounds 3 and 4**, no words were missing, indicating that the AI eventually reproduced the original text without errors.

---

**Comment by Nondzu** [comment]
@fairydreaming I found a possible issue with that, need to reconvert model again. see u soon

**Comment by Nondzu** [comment]
still the same issue, removed ignore_merges from llama-vocab.cpp and I did again a conversion and quant but no success. 'of' and `and` are still missing. 
https://github.com/Nondzu/llama.cpp/commit/9ec337849645604062f84ba31610ac1001d3dcb8
[log.txt](https://github.com/user-attachments/files/18549311/log.txt)

**Comment by fairydreaming** [comment]
@Nondzu OK, if it happens on Q8_0 then likely there's still some problem with my inference code as I didn't observe this behavior via API in OpenRouter. Thanks for testing!

**Comment by Kreijstal** [comment]
rip

**Comment by Kreijstal** [comment]
@ehartford maybe someone needs to open a new issue

**Comment by fairydreaming** [comment]
I recently returned to working on this (yeah, the model is obsolete at this point but I don't care) and found that the word-eating problem we observed was caused by special tokens from outside of the vocabulary (without any text representation) being produced by the model. In MiniMax-Text-01 vocab_size is 200064, but in tokenizer.json last defined token is ".chromium": 199999 and there are some added_tokens defined starting from 200000. When this problem occurs all logits are negative except for zeros at the end:

```
(Pdb) print(logits[0][0].shape)
torch.Size([200064])
(Pdb) print(logits[0][0][200000:])
tensor([-52.2500, -50.2500, -49.5000, -50.5000, -47.2500, -28.0000, -28.2500, -27.7500, -41.0000, -37.2500, -31.3750, -38.2500,
        -27.3750, -26.8750, -27.1250, -29.3750, -36.7500, -30.5000, -30.6250, -54.0000, -49.5000, -23.2500, -45.0000, -52.2500,
        -52.2500, -52.2500, -52.2500, -52.2500, -52.2500, -52.2500, -52.2500, -52.2500,   0.0000,   0.0000,   0.0000,   0.0000,
          0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,
          0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,
          0.0000,   0.0000,   0.0000,   0.0000])
```

if we omit these special tokens and see what's the largest logit:

```
(Pdb) print(logits[0][0][:200000].sort(descending=True)[0:10])
(tensor([ -2.1562, -18.1250, -18.2500,  ..., -52.7500, -53.0000, -53.0000]), tensor([   516,   9506,    812,  ...,  62649,  87979, 178727]))
(Pdb) print(logits[0][0][:200000].argsort(descending=True)[0:10])
tensor([  516,  9506,   812,   687, 14380,  1552,   311, 24263, 15383, 83202])
```

we see that it's token 516 `not` that should be sampled there. Unfortunately logit for this token is -2.1562, so zeros for the special tokens are sampled instead, which manifested as random words being eaten.

I guess the solution there is to somehow mask these special tokens during sampling. I checked embeddings for these tensors and:

```
(Pdb) print(self.lm_head.weight[200031])
tensor([ 0.0028,  0.0014, -0.0117,  ...,  0.0004,  0.0003, -0.0079], dtype=torch.bfloat16, requires_grad=True)
(Pdb) print(self.lm_head.weight[200032:])
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]], dtype=torch.bfloat16, requires_grad=True)
```

so all tokens starting from 200032 have zero-valued embeddings and will produce zero logit causing this issue.
